### DISCUSSION TITLE
Avoid building actions that won't be needed?

### DISCUSSION THREAD

**Discussion by rhodrin** [community/community#26043] [Actions]
<p>Lets say I have the following within a workflow:</p>
<pre><code class="lang-auto">    steps:
    - name: Checkout repo
      if: matrix.runner == 'runner1'
      uses: actions/checkout@v2

    - name: Do something
      if: matrix.runner == 'runner2'
      uses: some_name/some-action@master
</code></pre>
<p>where <code>runner1</code> and <code>runner2</code> refer to two different self hosted runners - we have some suitable, e.g., <code>runs-on: ${{ matrix.tags }}</code> at the start of the job which, lets say here, results in two jobs. So job 1 only needs <code>checkout</code> and job 2 only needs <code>some-action</code>. But at the start of both jobs both actions will be downloaded and built. If <code>some-action</code> relies on, e.g., docker which is not installed on <code>runner 1</code> the workflow fails, we get a <code>Build some_name/some-action@master</code>: <code>##[error]File not found: 'docker'</code>. But since this action won’t be needed on this machine we shouldn’t need to install docker just to get past this step.</p>
<p>Is there currently anyway around this without separating steps at a higher level?</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/rhodrin">@rhodrin</a>,</p>
<p>Two actions “owner/action1@v2” and “owner/action2@master”.<br>
If you want ‘action1’ only is downloaded and build to run on ‘runner1’, and ‘action2’ only on ‘runner2’, you can try to set your matrix like as this:</p>
<pre><code class="lang-yaml">jobs:
  job1:
    name: Test
    runs-on: ${{ matrix.runner }}
    strategy:
      matrix:
        include:
          - { runner: runner1, action_repos: 'owner/action1', action_ref: 'v2', s_name: 'step name to run action1' }
          - { runner: runner2, action_repos: 'owner/action2', action_ref: 'master', s_name: 'step name to run action2' }
    steps:
      - name: Checkout action code
        uses: actions/checkout@v2.3.2
        with:
          repository: ${{ matrix.action_repos }}
          ref: ${{ matrix.action_ref }}
          path: 'uses_action'

      - name: ${{ matrix.s_name }}
        uses: ./uses_action
</code></pre>
<p>In this way, the source code of the action will be downloaded via the checkout action. If the the action is docker action, the <strong>docker build</strong> will be executed in the step when running the action, it will not do the <strong>docker build</strong> before this step.</p>
<p>In your case, you also can set the matrix like this:</p>
<pre><code class="lang-yaml">jobs:
  job1:
    name: Test
    runs-on: ${{ matrix.runner }}
    strategy:
      matrix:
        include:
          - { runner: runner1 }
          - { runner: runner2, action_repos: 'some_name/some-action', action_ref: 'master', s_name: 'Do something' }
    steps:
      - name: Checkout repo
        if: matrix.runner == 'runner1'
        uses: actions/checkout@v2.3.2

      - name: Checkout action code
        if: matrix.runner == 'runner2'
        uses: actions/checkout@v2.3.2
        with:
          repository: ${{ matrix.action_repos }}
          ref: ${{ matrix.action_ref }}
          path: 'uses_action'

      - name: ${{ matrix.s_name }}
        if: matrix.runner == 'runner2'
        uses: ./uses_action
</code></pre>

**Comment by rhodrin** [comment]
<p><a class="mention" href="/brightran">@brightran</a></p>
<p>Thanks for the reply and suggestion. Modifying the <code>matrix</code>  to contain information about which runner requires which actions is not really an ideal solution in this instance for a variety of reasons, but I’ll have a think re. the best way forward.</p>
<p>Having a more ‘intelligent’ download and build phase based on <code>if</code> statements present would of course be the ideal solution.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/rhodrin">@rhodrin</a>,</p>
<blockquote>
<p>Modifying the <strong>matrix</strong> to contain information about which runner requires which actions is not really an ideal solution in this instance for a variety of reasons</p>
</blockquote>
<p>Yes, you’re right.<br>
When a runner requires multiple specific actions, this workaround indeed is not good enough.</p>
<p>In your case,  when the runner does not require the Docker action, the main thing you need to do is avoiding downloading and building the Docker action so that avoiding the error caused by the <strong>docker build</strong>.</p>
<p>For the JavaScript and Composite actions, they generally do not need to build before execute the actions. Normally it will not have much effect to download the source files of the actions, even if the runner may do not require the actions.</p>
<p>Alternatively, I have another workaround for your case that you do not need to change the configurations of the matrix. It uses the <strong>if</strong> statements to skip downloading, building and executing the Docker action.<br>
Theoretically, this workaround is similar to the previous one.</p>
<pre><code class="lang-yaml">jobs:
  job1:
    runs-on: ${{ matrix.runner }}
    strategy:
      matrix:
        runner: [runner1, runner2]
    steps:
      - name: Checkout repo
        if: matrix.runner == 'runner1'
        uses: actions/checkout@v2.3.2

      - name: Checkout action code
        if: matrix.runner == 'runner2'
        uses: actions/checkout@v2.3.2
        with:
          repository: some_name/docker-action
          ref: master
          path: 'docker-action'

      - name: Execute docker-action
        if: matrix.runner == 'runner2'
        uses: ./docker-action
</code></pre>

**Comment by rhodrin** [comment]
<p><a class="mention" href="/brightran">@brightran</a> Ok, I see - that could potentially work nicely, will try it out. Thanks again for the help.</p>
